### PR TITLE
Feature/zaink902/1076984273 edit endpoint

### PR DESCRIPTION
**Pull request by zainnk902**
### Administrative Info
Monday Board ID: 853628905
Make sure your branch name conforms to: `<feature/staging/hotfix/...>/[username]/[Monday Item ID]-[3-4 word description separated by dashes]`. Otherwise, please rename your branch and create a new PR.

### Changes
What changes did you make?
- Made the editItem endpoint so a user can edit the outDate and poundsRemaining of an existing batch in the database
- Prompted errors if outDate is before inDate or new weight is greater than the total pounds
- Gave a "Database is not connected" error if a user is trying to log in without the backend running
- Made sure the "Please fill in all fields" error popped up even if the backend wasn't running

### Testing
How did you confirm your changes worked? 
- We sent a PUT request to the editItem endpoint through Postman and saw if it made changes to the database object
- We ran the front end without the backend to test if the error messages still show up

### Confirmation of Change 

Screen Shot 2021-02-26 at 3 47 50 PM
Screen Shot 2021-02-26 at 3 52 47 PM

### CODE REVIEW

File: backend/routes/items.js [hunk 1]
```diff
@@ -125,4 +125,37 @@ router.post("/addItem", token_required, async (req, res, next) => {
...
+    const outDate = toUTCMidnight(new Date(req.body.outDate));
+    let batch = await Batch.findOne({ itemName, batchId });
+    
+    if (outDate < batch.inDate) {
+      return res.status(400).json({
+        fieldName: "outDate",
+        message: "Outdate is in the past",
+      });
+    }
+    if (weight > batch.poundsTotal) {
```
Comment (hahsu) [inline backend/routes/items.js]: I'm a little confused why we need to check for this. Is this saying that if there are only 50 pounds left to ship out, the user cannot edit the weight to be 80, or am I understanding this incorrectly? The user should be able to edit the number of pounds to any positive number, even if it is larger than the current total number of pounds to ship out.

File: backend/routes/items.js [hunk 2]
```diff
@@ -125,4 +125,37 @@ router.post("/addItem", token_required, async (req, res, next) => {
...
+      return res.status(400).json({ fieldName: "poundsRemaining", message: "Pounds remaining is greater than pounds total" })
+    }
+    
+    batch = await Batch.findOneAndUpdate({ itemName, batchId }, { outDate, poundsRemaining: weight });
```
Comment (hahsu) [inline backend/routes/items.js]: So I guess if the poundsTotal is bigger than it was before, we need to update that as well.

**Review by hahsu — CHANGES_REQUESTED** [review]

File: backend/routes/items.js (hunk 1 above)
Comment (lynguyxn) [reply backend/routes/items.js]: I was thinking that the poundsTotal is how much supply the user has of a batch of item. So if a user adds a batch of 50 pounds, the poundsTotal would be set to 50, and if they edit the batch whenever they send out a certain number of pounds, the poundsRemaining would be set to that weight they put. So the weight would be how much they still have to send out. If my understanding is correct, then the poundsRemaining can't be more than the poundsTotal. Correct me if I'm wrong!

File: backend/routes/items.js (hunk 2 above)
Comment (lynguyxn) [reply backend/routes/items.js:153]: The edit form on figma only has weight so I was following that. Should we add a field to edit the poundsTotal?

File: backend/routes/items.js (hunk 1 above)
Comment (hahsu) [reply backend/routes/items.js]: I think you are correct in that is the main usage of the edit button, however I was also thinking about the case the user accidentally slips up. For example, maybe they entered a batch of 50 pounds but they actually meant a batch of 500 pounds (missed a 0). I think we need to be able to allow for changing the total number of pounds for errors like that.

File: backend/routes/items.js [hunk 3]
```diff
@@ -125,4 +125,37 @@ router.post("/addItem", token_required, async (req, res, next) => {
   });
 });
 
+router.put("/editItem", token_required, async (req, res, next) => {
+  // authentication
+  try {
+    // req.user is the user id fetched from Middleware
+    const user = await User.findById(req.user.id);
+  } catch (e) {
+    console.log(e);
```
Comment (AnsonLee99) [inline backend/routes/items.js]: Let's stick with console.error() for error logging

File: backend/routes/items.js (hunk 1 above)
Comment (AnsonLee99) [reply backend/routes/items.js]: I agree with Hannah's point

Comment (lynguyxn) [reply backend/routes/items.js]: Fixed!

**Review by AnsonLee99 — CHANGES_REQUESTED** [review]
minor changes specified in comments

**Review by AnsonLee99 — APPROVED** [review]
lgtm!

**Review by hahsu — APPROVED** [review]
lgtm, thank you!

**Pull request merged**